docs: add event scanner specification document

yahgw-ai opened the pull request:
## Summary
- Added comprehensive specification document for the event scanner component
- Documented architecture, data structures, and implementation requirements
- Outlined testing approach and future considerations

## Test plan
- [x] All existing tests pass
- [x] No code changes, documentation only

yahgw-ai committed:
docs: add event scanner specification document

This commit introduces a detailed specification for the event scanner
component, outlining its architecture and implementation requirements.

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+### Primary Objectives
+
+- Scan blockchain for `RecipientRecieved` events from known reward distributor contracts
+- Aggregate daily outflow data for each distributor
```
We dont need to do any aggregation here - we just need to get and store the events

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+- Read distributor addresses and creation dates from persistent storage
+- Query blockchain events within specified date ranges
+- Parse and validate event data
+- Calculate daily outflow totals
```
out of scope

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+### Success Criteria
+
+- All `RecipientRecieved` events are captured for each distributor from creation date onward
+- Daily outflow totals match the sum of all events for that day
```
out of scope

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+- `withRetry` utility: Provides retry logic for RPC calls
+- `chunkBlockRange` utility: Splits large block ranges for efficient querying
+
+### Infrastructure Requirements
```
remove this section

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+
+### Input Flow
+
+1. Read distributor list from `store/distributors.json`
```
read data from the file manager - never mention file names in this spec

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+
+1. For each processing date:
+   - Calculate block range (previous day's block + 1 to current day's block)
+   - Query blockchain for `RecipientRecieved` events in range
```
Will need to use the chunking utility to do that

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+   - Calculate block range (previous day's block + 1 to current day's block)
+   - Query blockchain for `RecipientRecieved` events in range
+   - Parse event data and extract recipient, value, and transaction hash
+   - Aggregate events and calculate total outflow
```
dont calc outflows - remove mention of that everyhere

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+   * @returns Promise that resolves when distributor is processed
+   * @throws Error if distributor not found or processing fails
+   */
+  scanDistributor(distributorAddress: string): Promise<void>;
```
combine these into singke method with optional distributor

yahgwai commented on docs/specs/event-scanner.md:
````diff
@@ -0,0 +1,383 @@
+### Interface Definition
+
+```typescript
+interface EventScanner {
````
lets call this recipientrecievedscanner

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+
+   - Load distributor metadata (address, creation date, chain ID)
+   - Load existing outflow data or initialize empty structure
+   - Load master block numbers for date mapping
```
we can load the dates first outside this distributor loop

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+interface OutflowData {
+  metadata: {
+    chain_id: number; // Network chain ID
+    reward_distributor: string; // Distributor contract address
```
we need to record where we're up to - a last scanned block

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+    chain_id: number; // Network chain ID
+    reward_distributor: string; // Distributor contract address
+  };
+  outflows: {
```
instead of storing values, lets store raw data. lets store recipient recieved events in here, keyed by transactionhash:logindex. lets store raw data as well as parsed data were interested - recipient and value. we'll use n etheres interface for parsing

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+
+### Validation Rules
+
+1. **Input Validation**
```
ignore - the filmanger handles this

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+   - Block numbers file must contain required date ranges
+   - Dates must be in YYYY-MM-DD format
+
+2. **Event Validation**
```
just use an ethers interface to parse the event and it will handle this

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+   - Value must be valid uint256 (non-negative, within bounds)
+   - Transaction hash must be 66 characters (0x + 64 hex)
+
+3. **Output Validation**
```
filemanager does this

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+
+## 8. Implementation Requirements
+
+### Validation Rules
```
get rid of this section, see other comments

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+
+   - Validate all blockchain data before processing
+   - Use checksummed addresses for consistency
+   - Atomic file updates (write to temp, then rename)
```
ignore - not our remit

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+
+2. **Error Handling**
+
+   - Never partially update outflow data
```
ignore - not our remit

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+2. **Error Handling**
+
+   - Never partially update outflow data
+   - Preserve existing data on failure
```
get rid - not our remit

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+
+## 9. Test Data Requirements
+
+### Minimum Test Data
```
lets create a ticket to collect all the nova recipient recieved events. for testing with. we have the distributors in the test files already. the ticket should create a one if script to scan through all blocks looking for the events. dont create the ticket just write what the result should look like here and include that we need a ticket to get this data

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+
+### Data Sources
+
+- Mainnet Arbitrum reward distributor contracts
```
e'll use nova

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+### Data Sources
+
+- Mainnet Arbitrum reward distributor contracts
+- Historical blockchain data via archive node
```
dont need archive for this - and there is an rpc in other tests that we'll use the same as

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+    "2024-01-15": {
+      "block_number": 100123456,
+      "total_outflow_wei": "1500000000000000000",
+      "events": [
```
we've said elsewhere what this should be

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,383 @@
+  "outflows": {
+    "2024-01-15": {
+      "block_number": 100123456,
+      "total_outflow_wei": "1500000000000000000",
```
dont need this

yahgw-ai commented:
## Summary of addressed PR comments

I've addressed the following review comments:

1. **Comment #2163003557** - Removed aggregation throughout the spec. We now just store events without calculating totals.

2. **Comment #2163004034** - Already addressed (removed "Calculate daily outflow totals").

3. **Comment #2163004430** - Already addressed (removed "Daily outflow totals match...").

4. **Comment #2163008236** - Removed Infrastructure Requirements section.

5. **Comment #2163008564** - Updated all file path references to use FileManager instead of specific file names.

6. **Comment #2163009259** - Added mention of using chunkBlockRange utility in the Process Flow.

7. **Comment #2163012643** - Removed all mentions of calculating outflows.

8. **Comment #2163013725** - Combined scanAll() and scanDistributor() into a single scan() method with optional distributor parameter.

9. **Comment #2163014289** - Renamed EventScanner to RecipientRecievedScanner throughout the document.

10. **Comment #2163015633** - Moved loading of block numbers outside the distributor loop.

11. **Comment #2163019161** - Added last_scanned_block field to track scanning progress.

12. **Comment #2163024699** - Major restructure: Now storing raw event data keyed by transactionHash:logIndex instead of organizing by date. Events include both raw and parsed data.

The specification has been updated to reflect all these changes. 12 additional comments remain to be addressed.

yahgw-ai committed:
docs: address PR review comments on event scanner spec

- Remove aggregation logic, store events only
- Remove Infrastructure Requirements section
- Replace file paths with FileManager references
- Add chunkBlockRange utility mention
- Combine scan methods into single method with optional parameter
- Rename to RecipientRecievedScanner
- Move block numbers loading outside distributor loop
- Add last_scanned_block tracking field
- Restructure to store raw events keyed by txHash:logIndex

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,385 @@
+  value: string; // Decimal string representation of value
+}
+
+interface EventData {
```
Rename to receipientreceivedeventdata

yahgwai commented on docs/specs/event-scanner.md:
```diff
@@ -0,0 +1,385 @@
+# RecipientRecieved Scanner Component Technical Specification
```
Move the file to receipient-recieved-scanner.md

yahgw-ai committed:
docs: address remaining PR review comments on event scanner spec

- Remove references to FileManager-handled functionality (atomic updates, data preservation)
- Update test data requirements to specify Nova Arbitrum and standard RPC
- Add test data format specification
- Emphasize using ethers interface for event parsing

yahgw-ai committed:
docs: address final PR review comments

- Rename EventData interface to RecipientRecievedEventData
- Move spec file to recipient-recieved-scanner.md

yahgw-ai committed:
docs: add note about FileManager update ticket

Add note that a ticket is needed to update FileManager implementation
and specification to support the new RecipientRecievedEventData structure
